cli/azd/pkg/tools/azcli/subscriptions.go

```diff
@@ -0,0 +1,177 @@
+package azcli
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armsubscriptions"
+	azdinternal "github.com/azure/azure-dev/cli/azd/internal"
+	"github.com/azure/azure-dev/cli/azd/pkg/auth"
+	"github.com/azure/azure-dev/cli/azd/pkg/httputil"
+)
+
+// SubscriptionsService allows querying of subscriptions and tenants.
+type SubscriptionsService struct {
+	credentialProvider auth.MultiTenantCredentialProvider
+	userAgent          string
+	httpClient         httputil.HttpClient
+}
+
+func NewSubscriptionsService(
+	credentialProvider auth.MultiTenantCredentialProvider,
+	httpClient httputil.HttpClient) *SubscriptionsService {
+	return &SubscriptionsService{
+		userAgent:          azdinternal.MakeUserAgentString(""),
+		httpClient:         httpClient,
+		credentialProvider: credentialProvider,
+	}
+}
+
+func (ss *SubscriptionsService) createSubscriptionsClient(ctx context.Context, tenantId string) (*armsubscriptions.Client, error) {
+	options := clientOptionsBuilder(ss.httpClient, ss.userAgent).BuildArmClientOptions()
+	cred, err := ss.credentialProvider.GetTokenCredential(ctx, tenantId)
+	if err != nil {
+		return nil, err
+	}
+
+	client, err := armsubscriptions.NewClient(cred, options)
+	if err != nil {
+		return nil, fmt.Errorf("creating subscriptions client: %w", err)
+	}
+
+	return client, nil
+}
+
+func (ss *SubscriptionsService) createTenantsClient(ctx context.Context) (*armsubscriptions.TenantsClient, error) {
+	options := clientOptionsBuilder(ss.httpClient, ss.userAgent).BuildArmClientOptions()
+	// Use default home tenant, since tenants itself can be listed across tenants
+	cred, err := ss.credentialProvider.GetTokenCredential(ctx, "")
+	if err != nil {
+		return nil, err
+	}
+	client, err := armsubscriptions.NewTenantsClient(cred, options)
+	if err != nil {
+		return nil, fmt.Errorf("creating tenants client: %w", err)
+	}
+
+	return client, nil
+}
+
+func (s *SubscriptionsService) ListSubscriptions(ctx context.Context, tenantId string) ([]AzCliSubscriptionInfo, error) {
+	client, err := s.createSubscriptionsClient(ctx, tenantId)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Consider moving the creation of the `client` objects into the struct initializer `New` func or run a single func that only ever creates the client a single timer per instance of this service.

Reply:
@wbreza this is how `azcli` is implemented as well. For this particular case, we do need a `client` constructed with the given `tenantId` as part of method invocation, so we want to create it on method invocation.

Reply:
Agree with @wbreza 

You can keep the multiple sdk clients (one per tenant-id) allocated inside the `SubscriptionService` object.
When calling `createSubscriptionsClient` it should either return an existing client for that tenant, or create, persist and return the reference for the new client

Reply:
@wbreza @vhvb1989

Not going to address this for now. We would need to cache clients by-tenant to work. I don't believe this is an actual performance impact based on observation. The main thing that would slow us down is recreating the actual transport used by the clients. Otherwise, it's just a bunch of struct initialization to set up the pipeline. 

Note that we recreate clients on the fly in `azcli`, so we should revisit both when needed. And most importantly, `azcli` has a unrecommended practice of creating `httpClient` on the fly. I have a todo tracking to create this issue once things settle.